tensorzero-core/src/cache.rs

```diff
@@ -287,38 +301,35 @@ fn spawn_maybe_cache_write<T: Serialize + CacheOutput + Send + Sync + 'static>(
     });
 }
 
+/// Holds fields used to validate a `CacheData` before writing.
+/// We use this to skip writing certain 'bad' cache entries
+/// (e.g. when a tool call fails validation), while still allowing the
+/// inference itself to succeed and return a response to the user
+pub struct CacheValidationInfo {
+    // The `ToolCallConfig` for the top-level inference request, if present
+    // This is deliberately not part of the cache key - we only use it to
+    // skip writing certain cache entries.
+    pub tool_config: Option<ToolCallConfig>,
```

Comment:
we should eventually probably take the output schema for JSON functions here too. We can add a comment and issue for now and then deal with it later perhaps.